### ISSUE TITLE
firebase-tools used inside emulator callable cloud function deletes production data

### ISSUE THREAD

**Issue by maierson** [firebase/firebase-tools#2001]
### [REQUIRED] Environment info

**firebase-tools:**
7.14.0

**Platform:**
macOS

### [REQUIRED] Test case
See steps below:

### [REQUIRED] Steps to reproduce
Firebase callable function set up as here: https://firebase.google.com/docs/firestore/solutions/delete-collections.

The data present in the emulator was downloaded from production database and imported into the emulator according to the new import/export functionality available in 7.14.0

```ts
import * as tools from "firebase-tools";

export const deleteForever = functions
  .runWith({
    timeoutSeconds: 540,
    memory: "2GB",
  })
  .https.onCall((data, context) => {
     return tools.firestore
       .delete(path, {
        project: process.env.GCLOUD_PROJECT,
        recursive: true,
        yes: true,
        token: functions.config()?.fb?.token,
    })
    .then(() => path)
}
```
In the browser this is triggered like this:
```ts
firebase.functions.httpsCallable("deleteForever")(item)
```

### [REQUIRED] Expected behavior
Delete item and children in the local firestore emulator database.

### [REQUIRED] Actual behavior
Deletes item and children in the production firestore database. The item is no longer available in the production console.

Basically is there a way to set the `databaseURL` in firebase-tools to point to the local emulator database instead of the production one? (or any other method to force the cli to use the local emulated firestore instance instead of production)

Great job with the new import / export functionality btw. Works really well.

Thank you.

**Comment by samtstern** [comment]
@maierson wow that's a good catch!  I hope you didn't delete any important data due to this issue.

I will see what we can do to prevent this.  The issue is that `firebase-tools` uses the Firestore REST API so it's not monkey-patched in the same way.

**Comment by maierson** [comment]
@samtstern Nope :) We have a separate "production" db just for testing. Solving it will help our Cypress test runs via the emulator (it's how I ran into it).

Thanks for looking into this.

**Comment by samtstern** [comment]
Ok the good news is our network filtering does at lease notice these requests, I see this log:
```
⚠  Google API requested!
   - URL: "https://firestore.googleapis.com/v1beta1/projects/fir-dumpster/databases/(default)/documents:runQuery"
```

So I *hope* we can do something to just change the host.

**Comment by maierson** [comment]
👍
For whatever it's worth I DID delete a couple of online entities before I realized what is going on so if we didn't have the mock-production db I would have lost some data (barring no backups). I couldn't find this mentioned anywhere in the documentation so I imagine others will run into it. It's a pretty common use case for deleting entire entity trees.

**Comment by samtstern** [comment]
#2009 will make sure the emulators don't talk to prod and https://github.com/firebase/firebase-tools/issues/2010 will fix the fact that the local emulator does not actually behave correctly for these admin actions.

**Comment by maierson** [comment]
@samtstern Thank you for addressing this so promptly. Very impressive.

**Comment by samtstern** [comment]
@maierson happy to help!  I actually spent half a day implementing some crazy proxy thing before I realized `firebase-tools` already let me override this request with an environment variable.

I closed this issue because #2009 is merged which means prod won't be affected when it shouldn't.  However #2010 is the one to watch now because the Firestore emulator just doesn't respond well to the recursive delete calls.

**Comment by maierson** [comment]
Great. I've subscribed.

**Comment by samtstern** [comment]
@maierson let me know if everything works as expected for you in `7.15.1`

**Comment by maierson** [comment]
Hi Sam, Thank you for addressing this.

I now get a token error because functions.config() is an empty object. This didn't happen previously when the function was defaulting to production environment:

```js
⚠  Non-existent functions.config() value requested!
   - Path: "fb.token"
   - Learn more at https://firebase.google.com/docs/functions/local-emulator
⚠  Google API requested!
   - URL: "https://cloudresourcemanager.googleapis.com/v1/projects/<my-project>:testIamPermissions"
   - Be careful, this may be a production service.
```

I have set the credentials and they are being accessed:

```
 functions: Your GOOGLE_APPLICATION_CREDENTIALS environment variable points to /Users/.../<key-name>.json. Non-emulated services will access production using these credentials. Be careful!
```
but since `functions.config()` is an empty object I cannot execute this call (it just hangs and then times out)

```
firestore
      .delete(path, {
        project: process.env.GCLOUD_PROJECT,
        recursive: true,
        yes: true,
        token: functions.config()?.fb?.token,
      })
```

I googled this briefly but couldn't find an easy way out and I'm really backed up here in NYC because of the lockdown. If you have any quick suggestion on how to tackle this it would help a lot.

**Comment by samtstern** [comment]
@maierson ok I'm going to dig into this a bit more.

**Comment by samtstern** [comment]
Ok #2041 should fix this once and for all!

**Comment by maierson** [comment]
@samtstern Thank you. Will look into it as soon as it's pushed.

**Comment by samtstern** [comment]
The fix has been released in `7.16.0`

**Comment by maierson** [comment]
@samtstern  Thank you for this. I just gave it a try and I get `context.auth.uid` undefined when I attempt to validate the user before the deletion. In production this is set automatically. Do I need to do a specific setup for the emulator?

**Comment by samtstern** [comment]
Can you file a new issue with the code of your function and some instructions for me to reproduce?

**Comment by maierson** [comment]
Ok will do. It's actually the entire `auth` object on context that is undefined.

**Comment by samtstern** [comment]
@maierson might want to take a look at #2057 and #2058 ... there was an issue with `auth.uid` but `auth` in general should not be undefined.

**Comment by maierson** [comment]
@samstern I reviewed both of them. This is definitely related. My setup is almost the same as the one you document in #2058. However the entire `auth` object is missing (`console.log(context.auth) -> undefined`). This comment seems to imply the same: https://github.com/firebase/firebase-tools/issues/2057#issuecomment-603184472

Note that I can get to the token via `context.rawRequest.headers["x-callable-context-auth"]` but it doesn't seem right to use it instead of auth.

**Comment by samtstern** [comment]
@maierson thank you for looking into this.  Can you file an issue showing some more details about your test case?  If you can include the function code, logs, and the content of that header that would be amazing.

**Comment by maierson** [comment]
Done. #2059. Happy to try anything else that might help. 🙏
